src/ui.js

```diff
@@ -77,29 +77,62 @@ function handleFileSelect(map, evt) {
 
     modal.show();
 
-    let openFile = file => new Promise(resolve => {
+    let handleGpx = file => new Promise(resolve => {
         let reader = new FileReader();
-        reader.onload = () => resolve(reader.result);
+        reader.onload = () => {
+            parseGPX(reader.result)
+                .then(parsedTracks => {
+                    parsedTracks.forEach(track => {
+                        track.filename = file.name;
+                        tracks.push(track);
+                        map.addTrack(track);
+                        modal.addSuccess();
+                    });
+                })
+                .then(resolve);
+        };
         reader.readAsText(file, 'UTF-8');
     });
 
-    let loadFile = file => openFile(file)
-        .then(parseGPX)
-        .then(parsedTracks => {
-            parsedTracks.forEach(track => {
-                track.filename = file.name;
-                tracks.push(track);
-                console.info('Adding track:', track.name);
-                map.addTrack(track);
-                modal.addSuccess();
-            });
-        })
-        .catch(err => {
-            modal.addFailure({name: file.name, error: err});
+    let handleImage = file => new Promise(resolve => {
+        let image = new Image(file);
+
+        image.extractExifData().then((image) => {
+            if (!image) {
+                modal.addFailure({name: file.name, error: 'No geolocation data'});
+                return resolve();
+            }
+
+            map.addImage(image);
+            modal.addSuccess();
+            resolve();
         });
+    });
 
-    Promise.all(files.map(loadFile)).then(() => {
-        map.recenter();
+    let detectFileType = file => new Promise(resolve => {
+        let extension = file.name.split('.').pop().toLowerCase();
+
+        switch (extension) {
+            case 'gpx':
+                handleGpx(file)
+                    .then(resolve);
+                break;
+            case 'jpg':
+                handleImage(file)
+                    .then(resolve);
+                break;
+            default:
+                console.log(`File ${file.name} is an unsupported format.`);
+                modal.addFailure({name: file.name, error: 'Unsupported file format'});
+                resolve();
+        }
```

Comment:
```javascript
let detectFileType = async file => {
    let extension = ...
    try {
        switch (extension) {
            case 'gpx':
                return await handleGpx(file);
            case 'jpg':
            case 'jpeg':
                return await handleImage(file);
            default:
                throw 'Unsupported file format';
        }
    } catch (err) {
        modal.addFailure(…);
    }
}
```
Shouldn't this be called handleFile, as it does more than just detecting what the file type is, it also handles it based on said file type